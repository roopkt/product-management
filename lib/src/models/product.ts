export interface Product {
  id: string;
  name: string;
  price: number;
}

export function generateMockProduct(): Product {
  return { id: '2', name: 'Apple', price: 0 };
}

export function generateProduct({ name, price }): Product {
  return { name, price } as Product;
}
