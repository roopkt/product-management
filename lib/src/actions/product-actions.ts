import { Action } from '@ngrx/store';
import { Product } from '../models/product';

export enum ProductActionTypes {
  ADD = '[@scci-branding/product-management][Products] Add',
  ADD_SUCCESS = '[@scci-branding/product-management][Products] Add Success',
  ADD_ERROR = '[@scci-branding/product-management][Products] Add Error',
  VALIDATE_ADD_PRODUCT_FORM = '[@scci-branding/product-management][Products] Validate Form',
  LOAD = '[@scci-branding/product-management][Products] Load',
  LOAD_SUCCESS = '[@scci-branding/product-management][Products] Load Success',
}

export class Add implements Action {
  readonly type = ProductActionTypes.ADD;
  constructor(public payload: Product) { }
}

export class ValidateAddProduct implements Action {
  readonly type = ProductActionTypes.VALIDATE_ADD_PRODUCT_FORM;
  constructor(public payload: Product) { }
}

export class AddSuccess implements Action {
  readonly type = ProductActionTypes.ADD_SUCCESS;
  constructor(public payload: Product) { }
}

export class AddError implements Action {
  readonly type = ProductActionTypes.ADD_ERROR;
  constructor(public payload: string) { }
}

export class Load implements Action {
  readonly type = ProductActionTypes.LOAD;
  constructor() { }
}

export class LoadSuccess implements Action {
  readonly type = ProductActionTypes.LOAD_SUCCESS;
  constructor(public payload: Product[]) { }
}

export type ProductActions = Add
  | ValidateAddProduct
  | AddError
  | Load
  | LoadSuccess
  | AddSuccess;

