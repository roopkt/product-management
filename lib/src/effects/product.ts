import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { Scheduler } from 'rxjs/Scheduler';
import { async } from 'rxjs/scheduler/async';
import { empty } from 'rxjs/observable/empty';
import { of } from 'rxjs/observable/of';

import { ProductApi } from '../core/product.api';

import {
  Add,
  AddSuccess,
  AddError,
  LoadSuccess
} from '../actions/product-actions';

import { Product } from '../models/product';
import { ProductActionTypes } from '../actions/product-actions';
import { FeatureModuleConfig, FEATURE_MODULE_CONFIG } from '../config';
import {
  debounceTime,
  map,
  switchMap,
  skip,
  takeUntil,
  catchError,
} from 'rxjs/operators';

@Injectable()
export class ProductEffects {

  @Effect()
  addProduct$: Observable<Action> = this.actions$.pipe(
    ofType<Add>(ProductActionTypes.ADD),
    map(action => action.payload),
    switchMap(product => {
      return this.productApi
        .createProduct(product)
        .pipe(
        map((p: Product) => new AddSuccess(p)),
        catchError(err => of(new AddError(err))));
    })
  );

  @Effect()
  getAllProducts$: Observable<Action> = this.actions$.pipe(
    ofType<Add>(ProductActionTypes.LOAD),
    map(action => action.payload),
    switchMap(() => {
      return this.productApi
        .getAllProducts()
        .pipe(
        map(products => new LoadSuccess(products)),
        catchError(err => of(new AddError(err))));
    })
  );

  constructor(
    private actions$: Actions,
    private productApi: ProductApi
  ) {
  }
}
