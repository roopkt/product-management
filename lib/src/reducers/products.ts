import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import { ProductActions, ProductActionTypes } from '../actions/product-actions';
import { Product } from '../models/product';
import { createSelector } from '@ngrx/store';
import { MemoizedSelector } from '@ngrx/store/src/selector';
import { Dictionary, EntitySelectors } from '@ngrx/entity/src/models';

export interface State extends EntityState<Product> {
  loading: boolean;
  addProductForm: Product;
}

/**
 * createEntityAdapter creates many an object of helper
 * functions for single or multiple operations
 * against the dictionary of records. The configuration
 * object takes a record id selector function and
 * a sortComparer option which is set to a compare
 * function if the records are to be sorted.
 */
export const adapter: EntityAdapter<Product> = createEntityAdapter<Product>({
  selectId: (product: Product) => product.id,
  sortComparer: false,
});

/** getInitialState returns the default initial state
 * for the generated entity state. Initial state
 * additional properties can also be defined.
 */
export const initialState: State = adapter.getInitialState({
  loading: false,
  addProductForm: {} as Product
});

export function reducer(state = initialState, action: ProductActions): State {
  switch (action.type) {
    case ProductActionTypes.VALIDATE_ADD_PRODUCT_FORM: {
      return {
        ...state,
        addProductForm: action.payload
      };
    }
    case ProductActionTypes.ADD: {
      return {
        ...state,
        loading: true
      };
    }
    case ProductActionTypes.ADD_ERROR: {
      return {
        ...state,
        loading: false
      };
    }
    case ProductActionTypes.ADD_SUCCESS: {
      return {
        ...adapter.addOne(action.payload, state),
        loading: false
      };
    }
    case ProductActionTypes.LOAD_SUCCESS: {
      return {
        ...adapter.addMany(action.payload, state),
        loading: false
      };
    }
    default: {
      return state;
    }
  }
}

export const isLoading = (state: State) => state.loading;

export const getForm = (state: State) => state.addProductForm;

export const isNameValid: MemoizedSelector<State, any> = createSelector(
  getForm,
  (form: Product) => form.name != null
    && form.name.trim() !== ''
    && form.name.length > 3
    && form.name.length < 50);

export const isPriceValid = createSelector(
  getForm,
  (form: Product) => form.price != null
    && form.price > 0);

export const isFormValid = createSelector(
  isNameValid,
  isPriceValid,
  (isNameValidProp, isPriceValidProp) => isNameValidProp && isPriceValidProp
);

