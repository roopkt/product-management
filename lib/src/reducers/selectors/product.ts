import { createFeatureSelector, createSelector } from '@ngrx/store';
import { Product } from '../../models/product';
import { adapter, State, isFormValid, isNameValid, isPriceValid, getForm, initialState, isLoading } from './../products';
import { MemoizedSelector } from '@ngrx/store/src/selector';
import { Dictionary, EntitySelectors } from '@ngrx/entity/src/models';

export interface ProductsState {
  products: State;
}

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const getProductState: MemoizedSelector<
  any,
  ProductsState
  > = createFeatureSelector<ProductsState>('product-management-state');

/**
 * Every reducer module exports selector functions, however child reducers
 * have no knowledge of the overall state tree. To make them usable, we
 * need to make new selectors that wrap them.
 *
 * The createSelector function creates very efficient selectors that are memoized and
 * only recompute when arguments change. The created selectors can also be composed
 * together to select different pieces of state.
 */
export const getProductEntitiesState: MemoizedSelector<
  any,
  State
  > = createSelector(getProductState, state => (state) ? state.products : initialState);

/**
 * Adapters created with @ngrx/entity generate
 * commonly used selector functions including
 * getting all ids in the record set, a dictionary
 * of the records by id, an array of records and
 * the total number of records. This reducers boilerplate
 * in selecting records from the entity state.
 */
const entityFns = adapter.getSelectors(getProductEntitiesState);
export const {
    selectIds: getProductIds,
  selectAll: getAllProducts,
  selectTotal: getTotalProducts,
} = entityFns;

export const {
    selectEntities: getProductEntities,
}: EntitySelectors<Product, any> = entityFns;

export const isAddProductFormValid = createSelector(
  getProductEntitiesState,
  isFormValid);

export const isAddProductFormNameValid = createSelector(
  getProductEntitiesState,
  isNameValid);

export const isAddProductFormPriceValid = createSelector(
  getProductEntitiesState,
  isPriceValid);

export const addProductForm = createSelector(
  getProductEntitiesState,
  getForm);

export const isProductLoading = createSelector(
  getProductEntitiesState,
  isLoading);


