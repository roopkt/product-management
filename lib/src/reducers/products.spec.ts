import { Add, AddError, AddSuccess } from '../actions/product-actions';
import { generateMockProduct, Product } from '../models/product';
import { isLoading, reducer } from './products';
import * as fromProducts from './products';

describe('CenterDetail', () => {
    let initialState: fromProducts.State;

    describe('Action', () => {
        const product = generateMockProduct();

        initialState = {
            loading: false,
            ids: [product.id],
            entities: {
                [product.id]: product
            },
            addProductForm: null
        };

        describe('undefined action', () => {
            it('should return the default state', () => {
                const result = reducer(undefined, {} as any);
                expect(result).toMatchSnapshot();
            });
        });

        describe('ADD', () => {
            const action = new Add(product);
            it('Should return the state with loading true', () => {
                const result = reducer(initialState, action);
                expect(result).toMatchSnapshot();
            });
        });

        describe('ADD_SUCCESS', () => {
            const action = new AddSuccess(product);
            it('Should return the state with loading false', () => {
                const result = reducer(initialState, action);
                expect(result).toMatchSnapshot();
            });
        });

        describe('ADD_ERROR', () => {
            const action = new AddError('error');
            it('Should return the state with loading false', () => {
                const result = reducer(initialState, action);
                expect(result).toMatchSnapshot();
            });
        });
    });

    describe('Query', () => {
        describe('isLoading', () => {
            it('should return loading status', () => {
                const result = isLoading(initialState);
                expect(result).toMatchSnapshot();
            });
        });

        describe('isFormValid', () => {
            it('should return false', () => {
                const product = {
                    name: '',
                    price: 0
                } as Product;

                initialState = {
                    loading: false,
                    ids: [product.id],
                    entities: {
                        [product.id]: product
                    },
                    addProductForm: product
                };
                const result = isLoading(initialState);
                expect(result).toMatchSnapshot();
            });

            it('should return true', () => {
                const product = generateMockProduct();

                initialState = {
                    loading: false,
                    ids: [product.id],
                    entities: {
                        [product.id]: product
                    },
                    addProductForm: product
                };
                const result = isLoading(initialState);
                expect(result).toMatchSnapshot();
            });
        });
    });
});
