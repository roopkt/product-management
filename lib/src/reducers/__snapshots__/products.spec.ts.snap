// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CenterDetail Action ADD Should return the state with loading true 1`] = `
Object {
  "addProductForm": null,
  "entities": Object {
    "2": Object {
      "id": "2",
      "name": "Apple",
      "price": 0,
    },
  },
  "ids": Array [
    "2",
  ],
  "loading": true,
}
`;

exports[`CenterDetail Action ADD_ERROR Should return the state with loading false 1`] = `
Object {
  "addProductForm": null,
  "entities": Object {
    "2": Object {
      "id": "2",
      "name": "Apple",
      "price": 0,
    },
  },
  "ids": Array [
    "2",
  ],
  "loading": false,
}
`;

exports[`CenterDetail Action ADD_SUCCESS Should return the state with loading false 1`] = `
Object {
  "addProductForm": null,
  "entities": Object {
    "2": Object {
      "id": "2",
      "name": "Apple",
      "price": 0,
    },
  },
  "ids": Array [
    "2",
  ],
  "loading": false,
}
`;

exports[`CenterDetail Action undefined action should return the default state 1`] = `
Object {
  "addProductForm": Object {},
  "entities": Object {},
  "ids": Array [],
  "loading": false,
}
`;

exports[`CenterDetail Query isFormValid should return false 1`] = `false`;

exports[`CenterDetail Query isFormValid should return true 1`] = `false`;

exports[`CenterDetail Query isLoading should return loading status 1`] = `false`;
